### DISCUSSION TITLE
Wi-Fi interaction with timer interrupts?

### DISCUSSION THREAD

**Discussion by timg11** [micropython/micropython#10721] [ESP32]
I'm relatively new to the ESP32 platform with Micropython, and I'm using timer interrupts for the first time.
I have "MicroPython v1.15 on 2021-04-18; ESP32 module with ESP32" on the "WROOM" board.

The overall objective is measuring relatively slow analog signals and sending data by Wi-Fi to a simple TCP socket with `s.send(sendstring)`. I have a 50 mS timer, and the ISR performs the A/D conversions, and (in this early test phase) simply sets pins with LEDs on or off based on a fixed threshold of the A/D values. I'm driving the A/D inputs with a test signal (square wave approximately 1 to 10 Hz). 
The Wi-Fi send is in the main `while True` loop that has a `time.sleep_ms(1000)` and the Wi-Fi sends.  The main loop also reads temperature from two DS1820 sensors.
```py
        if DS1820_present:
            ds_sensor.convert_temp()
```
The LEDs should follow the test signal. They do, if I send a single WiFi message of about 50 bytes per second (main loop iteration). The problem occurs when I uncomment two other Wi-Fi sends resulting in a total of about 1k bytes per main loop iteration. The LEDs become erratic and don't follow the input signal. Possibly the Wi-Fi routines are blocking or otherwise affecting the timer interrupt service routines? 

I've read the [micropython docs for the ESP32](https://docs.micropython.org/en/latest/esp32/tutorial/intro.html) but there isn't much about interrupts and interactions with Wi-Fi. Where should I look for more information and guidance on ensuring interrupts are serviced on time while using Wi-Fi? 

The timer is configured as:
```py
    timer.init(period=50, mode=machine.Timer.PERIODIC, callback=handleInterrupt)      # period in ms. 
```

The handler is:
```py
def handleInterrupt(timer):
    global interruptCounter

    interruptCounter += 1
    for adcindex  in adcgroup:           # read these A/D converters
        adcrawvals[adcindex] = ADCarray[adcindex].read()

  #  snip -  based on values of two AD measurements, turn GPIO pins driving LEDs on or off
```

**Comment by robert-hh** [comment]
Which analog ports do you use? ADC2 was added in Jan 22. It was not included for a long time because of interference with the WiFi usage. But I do not recall the details.

**Reply by timg11** [reply]
The ADC definitions are below. Based on the [ESP32 Micropython documentation](https://docs.micropython.org/en/latest/esp32/quickref.html#adc-analog-to-digital-conversion), "Block 2" ADCs are not available with WI-Fi enabled, so these are the ones that remain in Block 1. 
```py
    ADCarray = [ADC(Pin(36)),   # PS1  AD0
                ADC(Pin(39)),   # PS2  AD3 
                ADC(Pin(35)),   # PS3  AD7
                ADC(Pin(34)),   # Inducer CT   AD6
                ADC(Pin(32)),   # Fault Sensor   AD4
                ADC(Pin(33))]   # Status Sensor   AD5
```

I have tried concatenating all data into a longer string with single call to `s.send(sendstring)`, rather than calling it multiple times for each string. That seems to help. I'm still curious as to the underlying interactions
